src/System.Security.Cryptography.Algorithms/tests/HKDFTests.cs

```diff
@@ -0,0 +1,611 @@
+        [InlineData(0)]
+        [InlineData(19)]
+        public void Rfc5869ExtractSpanPrkTooShort(int prkSize)
+        {
+            byte[] prk = new byte[prkSize];
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentException>(() => HKDF.Extract(HashAlgorithmName.SHA1, ikm, salt, prk));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractSpanPrkTooLong()
+        {
+            byte[] prk = new byte[24];
+
+            for (int i = 0; i < 4; i++)
+            {
+                prk[20 + i] = (byte)(i + 5);
+            }
+
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Equal(20, HKDF.Extract(HashAlgorithmName.SHA1, ikm, salt, prk));
+            Assert.Equal("A3CBF4A40F51A53E046F07397E52DF9286AE93A2".HexToByteArray(), prk.Take(20).ToArray());
+
+            for (int i = 0; i < 4; i++)
+            {
+                // ensure we didn't modify anything further
+                Assert.Equal((byte)(i + 5), prk[20 + i]);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCases))]
+        public void Rfc5869ExpandByteArrayTests(Rfc5869TestCase test)
+        {
+            byte[] okm = HKDF.Expand(test.Hash, test.Prk, test.Okm.Length, test.Info);
+            Assert.Equal(test.Okm, okm);
+        }
+
+        [Fact]
+        public void Rfc5869ExpandByteArrayDefaultHash()
+        {
+            byte[] prk = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Expand(default(HashAlgorithmName), prk, 20, null));
+        }
+
+        [Fact]
+        public void Rfc5869ExpandByteArrayNonsensicalHash()
+        {
+            byte[] prk = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Expand(new HashAlgorithmName("foo"), prk, 20, null));
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCases))]
+        public void Rfc5869ExpandByteArrayTamperPrkTests(Rfc5869TestCase test)
+        {
+            test.Prk[0] ^= 1;
+            byte[] okm = HKDF.Expand(test.Hash, test.Prk, test.Okm.Length, test.Info);
+            Assert.NotEqual(test.Okm, okm);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        [InlineData(19)]
+        public void Rfc5869ExpandByteArrayPrkTooSmall(int prkSize)
+        {
+            byte[] prk = new byte[prkSize];
+            Assert.Throws<ArgumentException>(() => HKDF.Expand(HashAlgorithmName.SHA1, prk, 17, Array.Empty<byte>()));
+        }
+
+        [Fact]
+        public void Rfc5869ExpandByteArrayOkmMaxSize()
+        {
+            byte[] prk = new byte[20];
+
+            // Does not throw
+            byte[] okm = HKDF.Expand(HashAlgorithmName.SHA1, prk, 20 * 255, Array.Empty<byte>());
+            Assert.Equal(20 * 255, okm.Length);
+        }
+
+        [Fact]
+        public void Rfc5869ExpandByteArrayOkmMaxSizePlusOne()
```

Comment:
8421505 is an interesting test case, too.  (ends up being -2147483521 after the multiplication)

Reply:
tests don't show overflow but it's good test either way